bin/varnishd/cache/cache_conn_pool.c

```diff
@@ -249,17 +259,39 @@ VCP_Rel(struct conn_pool **cpp)
 		(void)shutdown(pfd->fd, SHUT_RDWR);
 		cp->n_kill++;
 	}
-	while (cp->n_kill) {
-		Lck_Unlock(&cp->mtx);
-		(void)usleep(20000);
+
+	ALLOC_OBJ(dcp, DYING_CP_MAGIC);
+	dcp->cp = cp;
+	VTAILQ_INSERT_TAIL(&dying_cps, dcp, list);
+	Lck_Unlock(&cp->mtx);
+}
+
+void
+VCP_RelPoll(void)
+{
+	struct dying_cp *dcp, *dcp2;
+	struct conn_pool *cp;
+
+	VTAILQ_FOREACH_SAFE(dcp, &dying_cps, list, dcp2) {
+		cp = dcp->cp;
+		CHECK_OBJ_NOTNULL(cp, CONN_POOL_MAGIC);
+
 		Lck_Lock(&cp->mtx);
+		while (cp->n_kill) {
+			Lck_Unlock(&cp->mtx);
+			(void)usleep(20000);
```

Comment:
why not `return` instead? This is called from `cli_cb_before` hook, you can not afford to block here or you make it synchronous again. :smile:

Reply:
Marking this as duplicate. See below.